casper/src/test/scala/coop/rchain/casper/batch2/BlockReceiverEffectsSpec.scala

```diff
@@ -0,0 +1,234 @@
+    with ArgumentMatchersSugar {
+  implicit val logEff: Log[Task] = Log.log[Task]
+
+  it should "pass correct block to output stream with calling effectful components" in
+    withEnv[Task]("root") {
+      case (incomingQueue, _, outStream, bs, br, bds) =>
+        for {
+          block   <- Sync[Task].delay(makeBlock())
+          _       <- incomingQueue.enqueue1(block)
+          outList <- outStream.take(1).compile.toList
+        } yield {
+          bs.put(Seq((block.blockHash, block))) wasCalled once
+          bs.contains(Seq(block.blockHash)) wasCalled once
+          br.ackReceived(block.blockHash) wasCalled once
+          dagStorageWasNotModified(bds)
+          outList shouldBe List(block.blockHash)
+        }
+    }
+
+  // Provided to BlockReceiver shard name ("test") is differ from block's shard name ("root" by default)
+  // So block should be rejected and output stream should never take block
+  it should "discard block with invalid shard name" in withEnv[Task]("test") {
+    case (incomingQueue, _, outStream, bs, br, bds) =>
+      for {
+        block <- Sync[Task].delay(makeBlock())
+        _     <- incomingQueue.enqueue1(block)
+      } yield {
+        bs.put(*) wasNever called
+        bs.contains(*) wasNever called
+        br.ackReceived(*) wasNever called
+        dagStorageWasNotModified(bds)
+        outStream should notEmit
+      }
+  }
+
+  it should "discard block with invalid block hash" in withEnv[Task]("root") {
+    case (incomingQueue, _, outStream, bs, br, bds) =>
+      for {
+        block <- Sync[Task].delay(makeBlock().copy(blockHash = "abc".unsafeHexToByteString))
+        _     <- incomingQueue.enqueue1(block)
+      } yield {
+        bs.put(*) wasNever called
+        bs.contains(*) wasNever called
+        br.ackReceived(*) wasNever called
+        dagStorageWasNotModified(bds)
+        outStream should notEmit
+      }
+  }
+
+  it should "discard block with invalid signature" in withEnv[Task]("root") {
+    case (incomingQueue, _, outStream, bs, br, bds) =>
+      for {
+        block <- Sync[Task].delay(makeBlock().copy(sig = "abc".unsafeHexToByteString))
+        _     <- incomingQueue.enqueue1(block)
+      } yield {
+        bs.put(*) wasNever called
+        bs.contains(*) wasNever called
+        br.ackReceived(*) wasNever called
+        dagStorageWasNotModified(bds)
+        outStream should notEmit
+      }
+  }
+
+  it should "discard known block" in withEnv[Task]("root") {
+    case (incomingQueue, _, outStream, bs, br, bds) =>
+      for {
+        block <- addBlock[Task](bs)
+        _     <- incomingQueue.enqueue1(block)
+      } yield {
+        bs.put(Seq((block.blockHash, block))) wasCalled once
+        bs.contains(*) wasNever called
+        br.ackReceived(*) wasNever called
+        dagStorageWasNotModified(bds)
+        outStream should notEmit
+      }
+  }
+
+  it should "pass to output blocks with resolved dependencies" in withEnv[Task]("root") {
```

Comment:
There is problem with the test as we can see in [log](https://github.com/rchain/rchain/runs/7280313776?check_suite_focus=true#step:8:59).
When running locally, the test passes in most cases, but fails when run on github.
I have been debugging it for 2 days, but have not yet reached the results, since the bug does not appear regularly. To detect it, I wrote a loop in the shell to run `BlockReceiver` tests, but `sbt` is initialized every time it runs, which is very long.

```scala
// The arrows show the lines of code that the log refers to.
...
for {
  // Save block to block store, resolve parents to request
  _ <- BlockStore[F].put(block)
  parents <- block.justifications  // <-
              .traverse { hash =>
                BlockStore[F].contains(hash).not.map((hash, _))  // <-
              }
  pendingRequests <- state.modify(_.endStored(block.blockHash, parents))  // <-

  // Notify BlockRetriever of finished validation of block
  _ <- BlockRetriever[F].ackReceived(block.blockHash)  // <- 

  // Send request for missing dependencies
  _ <- requestMissingDependencies(pendingRequests)

  // Check if block have all dependencies in the DAG
  dag <- BlockDagStorage[F].getRepresentation
  hasAllDeps = pendingRequests.isEmpty &&
    block.justifications.forall(dag.contains)

  _ <- receiverOutputQueue.enqueue1(block.blockHash).whenA(hasAllDeps)
} yield ()
...
```

Judging by the fact that the problem does not appear regularly, it is related to concurrent code execution, and the NullPointerException error says that some object we are accessing has been destroyed.
Judging by the log, either the block or the mocks are destroyed. I can't say for sure yet.

Reply:
Hm, this can be a big issue. Did you find similar cases already reported with Mockito library?

Reply:
The problem could something similar as described here.
https://www.javafixing.com/2022/03/fixed-mockito-nullpointerexception-when.html

Reply:
I resolved the problem. The problem was that the `admitHash` method was not implemented, although it was called in some scenarios. The Mockito library did not indicate this, but only threw a `NullPointerException.` I was able to spot the problem by manually implementing mocks and noticing that a `NotImplementedError` was fired on the `admitHash` method that had an implementation of `???`.

Reply:
Awesome catch! It would be nice that Mockito can tell which method is called in that case.